File: Deque/10866.cpp
#include <iostream>
using namespace std;

const int MX = 1000005;
int dat[2 * MX + 1];
int head = MX, tail = MX;

void push_front(int x) {
	dat[--head] = x;
}

void push_back(int x) {
	dat[tail++] = x;
}

void pop_front() {
	head++;
}

void pop_back() {
	tail--;
}

int front() {
	return dat[head];
}

int back() {
	return dat[tail - 1];
}

int main(void)
{
	ios::sync_with_stdio(0);
	cin.tie(0);
	int n;
	cin >> n;
	while (n--) {
		string s;
		cin >> s;
		if (s == "push_front") {
			int a;
			cin >> a;
			push_front(a);
		}
		else if (s == "push_back") {
			int a;
			cin >> a;
			push_back(a);
		}
		else if (s == "pop_front") {
			if (tail - head == 0)
				cout << "-1" << '\n';
			else {
				cout << front() << '\n';
				pop_front();
			}
		}
		else if (s == "pop_back") {
			if (tail - head == 0)
				cout << "-1" << '\n';
			else {
				cout << back() << '\n';
				pop_back();
			}
		}
		else if (s == "size") {
			cout << tail - head << '\n';
		}
		else if (s == "empty") {
			if (tail - head == 0)
				cout << "1" << '\n';
			else
				cout << "0" << '\n';
		}
		else if (s == "front") {
			if (tail - head == 0)
				cout << "-1" << '\n';
			else
				cout << front() << '\n';
		}
		else if (s == "back") {
			if (tail - head == 0)
				cout << "-1" << '\n';
			else 
				cout << back() << '\n';
		}
	}
}

File: Deque/10866.kt
package practice

import java.util.*

fun main() = with(System.out.bufferedWriter()) {
    val br = System.`in`.bufferedReader()
    val n = br.readLine().toInt()
    val sb = StringBuilder()
    val deq: Deque<Int> = LinkedList<Int>()
    for (i in 0 until n) {
        val input = br.readLine().split(" ")
        when (input[0]) {
            "push_front" -> {
                deq.offerFirst(input[1].toInt())
            }
            "push_back" -> {
                deq.offerLast(input[1].toInt())
            }
            "pop_front" -> {
                if (deq.isNotEmpty()) {
                    sb.append(deq.pollFirst()).append('\n')
                } else {
                    sb.append(-1).append('\n')
                }
            }
            "pop_back" -> {
                if (deq.isNotEmpty()) {
                    sb.append(deq.pollLast()).append('\n')
                } else {
                    sb.append(-1).append('\n')
                }
            }
            "size" -> {
                sb.append(deq.size).append('\n')
            }
            "empty" -> {
                if (deq.isNotEmpty()) {
                    sb.append(0).append('\n')
                } else {
                    sb.append(1).append('\n')
                }
            }
            "front" -> {
                if (deq.isNotEmpty()) {
                    sb.append(deq.peekFirst()).append('\n')
                } else {
                    sb.append(-1).append('\n')
                }
            }
            "back" -> {
                if (deq.isNotEmpty()) {
                    sb.append(deq.peekLast()).append('\n')
                } else {
                    sb.append(-1).append('\n')
                }
            }
        }
    }
    println(sb.toString())
}
